### PR TITLE
Feature/24 modify single wallet

### PR DESCRIPTION
**Pull request by krystian14d**

### CODE REVIEW

File: src/main/java/pl/byczazagroda/trackexpensesappbackend/dto/EditWalletDto.java [hunk 1]
```diff
@@ -0,0 +1,11 @@
+package pl.byczazagroda.trackexpensesappbackend.dto;
+
+import javax.validation.constraints.Min;
+import javax.validation.constraints.NotBlank;
+import javax.validation.constraints.NotEmpty;
+import javax.validation.constraints.NotNull;
+import javax.validation.constraints.Pattern;
+import javax.validation.constraints.Size;
+
+public record EditWalletDto(@NotNull @NotEmpty @Min(value = 1) long id, @NotBlank @NotEmpty @Size(max = 20) @Pattern(regexp = "[a-z A-Z]+") String name) {
```
Comment (pyanik) [inline src/main/java/pl/byczazagroda/trackexpensesappbackend/dto/EditWalletDto.java:10]: Is @NotNull annotation needed? id field can not be null as a primite type.

Comment (dawciobiel) [inline src/main/java/pl/byczazagroda/trackexpensesappbackend/dto/EditWalletDto.java:10]: Just like @pyanik said primitive variable had to have value, so no point to type validations (NotNull, NotEmpty) on "id" parameter.

**Review by dawciobiel — CHANGES_REQUESTED** [review]
Please implements changes.

File: src/main/java/pl/byczazagroda/trackexpensesappbackend/dto/EditWalletDto.java (hunk 1 above)
Comment (dawciobiel) [reply src/main/java/pl/byczazagroda/trackexpensesappbackend/dto/EditWalletDto.java:10]: **Update:**
**Please do not remove validation annotations from this argument.** Change type of it from `primitive long` to `objective Long`. So in this case we will be sure it's validation controlling flow and we are catching this type of exception in our proper place.

Comment (krystian14d) [reply src/main/java/pl/byczazagroda/trackexpensesappbackend/dto/EditWalletDto.java:10]: Done - changed to objective Long.

File: src/main/java/pl/byczazagroda/trackexpensesappbackend/service/WalletService.java [hunk 2]
```diff
@@ -1,4 +1,9 @@
 package pl.byczazagroda.trackexpensesappbackend.service;
 
+import pl.byczazagroda.trackexpensesappbackend.dto.EditWalletDto;
+import pl.byczazagroda.trackexpensesappbackend.dto.WalletDTO;
+
 public interface WalletService {
+
+    WalletDTO updateWallet(EditWalletDto walletToEdit);
```
[the remark this answers is not in the document]
Comment (krystian14d) [reply src/main/java/pl/byczazagroda/trackexpensesappbackend/service/WalletService.java:8]: Changed to "UpdateWalletDto" convention.

File: src/main/java/pl/byczazagroda/trackexpensesappbackend/dto/UpdateWalletDto.java [hunk 3]
```diff
@@ -7,5 +7,5 @@
 import javax.validation.constraints.Pattern;
 import javax.validation.constraints.Size;
 
-public record EditWalletDto(@NotNull @NotEmpty @Min(value = 1) long id, @NotBlank @NotEmpty @Size(max = 20) @Pattern(regexp = "[a-z A-Z]+") String name) {
+public record UpdateWalletDto(@NotNull @NotEmpty @Min(value = 1) Long id, @NotBlank @NotEmpty @Size(max = 20) @Pattern(regexp = "[a-z A-Z]+") String name) {
```
Comment (dawciobiel) [inline src/main/java/pl/byczazagroda/trackexpensesappbackend/dto/UpdateWalletDto.java:10]: Not `UpdateWalletDto` but `UpdateWalletDTO`

File: src/main/java/pl/byczazagroda/trackexpensesappbackend/service/WalletServiceImpl.java [hunk 4]
```diff
@@ -22,7 +22,7 @@ public class WalletServiceImpl implements WalletService {
 
     @Override
     @Transactional
-    public WalletDTO updateWallet(EditWalletDto walletToEdit) throws WalletNotFoundException {
+    public WalletDTO updateWallet(UpdateWalletDto walletToEdit) throws WalletNotFoundException {
```
Comment (dawciobiel) [inline src/main/java/pl/byczazagroda/trackexpensesappbackend/service/WalletServiceImpl.java:25]: not `UpdateWalletDTO walletToUpdate`
but `UpdateWalletDTO walletToUpdate`

**Review by dawciobiel — CHANGES_REQUESTED** [review]
Please do code changes.

File: src/main/java/pl/byczazagroda/trackexpensesappbackend/dto/UpdateWalletDto.java (hunk 3 above)
Comment (krystian14d) [reply src/main/java/pl/byczazagroda/trackexpensesappbackend/dto/UpdateWalletDto.java]: done

File: src/main/java/pl/byczazagroda/trackexpensesappbackend/service/WalletServiceImpl.java (hunk 4 above)
Comment (krystian14d) [reply src/main/java/pl/byczazagroda/trackexpensesappbackend/service/WalletServiceImpl.java]: done

File: src/main/java/pl/byczazagroda/trackexpensesappbackend/model/Wallet.java [hunk 5]
```diff
@@ -45,8 +45,8 @@ public class Wallet implements Serializable {
 
     private Instant creationDate;
 
-    public Wallet(String name) {
+    public Wallet(String name, Instant creationDate) {
```
Comment (pyanik) [inline src/main/java/pl/byczazagroda/trackexpensesappbackend/model/Wallet.java:48]: We should stick with the previous, one parameter constructor (only String name argument).
We do not want the client to send us creation date value via request while creating or updating wallet.
Creation Date field should be also generated automatically by the server (via `Instant.now()` method.

File: src/main/java/pl/byczazagroda/trackexpensesappbackend/service/WalletService.java [hunk 6]
```diff
@@ -1,4 +1,9 @@
 package pl.byczazagroda.trackexpensesappbackend.service;
 
+import pl.byczazagroda.trackexpensesappbackend.dto.UpdateWalletDTO;
+import pl.byczazagroda.trackexpensesappbackend.dto.WalletDTO;
+
 public interface WalletService {
+
+    WalletDTO updateWallet(UpdateWalletDTO walletToEdit);
```
Comment (pyanik) [inline src/main/java/pl/byczazagroda/trackexpensesappbackend/service/WalletService.java:8]: We can change name of the argument to `walletToUpdate` as we do in `WalletServiceImpl.java`

File: src/test/java/pl/byczazagroda/trackexpensesappbackend/service/WalletServiceImplTest.java [hunk 7]
```diff
@@ -1,7 +1,67 @@
...
 class WalletServiceImplTest {
 
+    @MockBean
+    private WalletRepository walletRepository;
+
+    @MockBean
+    private WalletModelMapper mapper;
+
+    @Autowired
+    private WalletServiceImpl underTest;
```
Comment (pyanik) [inline src/test/java/pl/byczazagroda/trackexpensesappbackend/service/WalletServiceImplTest.java:39]: We can change the type to WalletService in line with dependency inversion principle

File: src/main/java/pl/byczazagroda/trackexpensesappbackend/model/Wallet.java (hunk 5 above)
Comment (krystian14d) [reply src/main/java/pl/byczazagroda/trackexpensesappbackend/model/Wallet.java:48]: Ok but it was told to stick to convention used in task 23, already merged to branch develop. There is used convention with two-args constructor. So why should I roll back to one-arg constructor? I see some misaligment here... Can you Please clarify it?

File: src/test/java/pl/byczazagroda/trackexpensesappbackend/service/WalletServiceImplTest.java [hunk 8]
```diff
@@ -1,7 +1,67 @@
...
+
+    @Test
+    void itShouldUpdateWalletName() {
+        //GIVEN
+        UpdateWalletDTO updateWalletDto = returnEditWalletDto();
```
Comment (pyanik) [inline src/test/java/pl/byczazagroda/trackexpensesappbackend/service/WalletServiceImplTest.java:44]: We don't need to extract method here. Let's just write:
UpdateWalletDTO updateWalletDto = new UpdateWalletDTO(1L, "walletName");

File: src/main/java/pl/byczazagroda/trackexpensesappbackend/model/Wallet.java (hunk 5 above)
Comment (pyanik) [reply src/main/java/pl/byczazagroda/trackexpensesappbackend/model/Wallet.java:48]: Take  a look at `Wallet.java` in develop branch. There is only one-parameter constructor with String name argument

image

File: src/test/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletControllerTest.java [hunk 9]
```diff
@@ -1,7 +1,240 @@
...
+                new WalletDTO(
+                        editWalletDto.getId(),
+                        editWalletDto.getName(),
+                        timeCreated
+                )
+        );
+
+        //WHEN
+        ResultActions editResultActions = mockMvc.perform(
+                patch("/api/wallet/edit")
```
Comment (pyanik) [inline src/test/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletControllerTest.java:64]: In Controller layer we defined put rest api method with `/api/wallet/update` endpoint but in `WalletControllerTest.java` we test patch method with `/api/wallet/edit` endpoint. It should be corrected

File: src/main/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletController.java [hunk 10]
```diff
@@ -14,4 +26,19 @@
 public class WalletController {
 
     private final WalletService walletService;
+
+    @PutMapping("/update")
+    public ResponseEntity<WalletDTO> updateWallet(UpdateWalletDTO updateWalletDto) {
+        WalletDTO walletDTO = walletService.updateWallet(updateWalletDto);
```
Comment (pyanik) [inline src/main/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletController.java:32]: `@RequestBody` and `@Valid` annotations missing

File: src/main/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletController.java [hunk 11]
```diff
@@ -14,4 +26,19 @@
 public class WalletController {
 
     private final WalletService walletService;
+
+    @PutMapping("/update")
```
Comment (pyanik) [inline src/main/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletController.java:30]: Maybe we should pass `id` of the wallet to update by the endpoint? so `/update//{id}` or even just `{id}`

File: src/main/java/pl/byczazagroda/trackexpensesappbackend/model/Wallet.java (hunk 5 above)
Comment (krystian14d) [reply src/main/java/pl/byczazagroda/trackexpensesappbackend/model/Wallet.java:48]: OK, updated.

File: src/main/java/pl/byczazagroda/trackexpensesappbackend/service/WalletService.java (hunk 6 above)
Comment (krystian14d) [reply src/main/java/pl/byczazagroda/trackexpensesappbackend/service/WalletService.java:8]: Updated.

File: src/test/java/pl/byczazagroda/trackexpensesappbackend/service/WalletServiceImplTest.java (hunk 8 above)
Comment (krystian14d) [reply src/test/java/pl/byczazagroda/trackexpensesappbackend/service/WalletServiceImplTest.java:44]: Updated.

File: src/test/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletControllerTest.java (hunk 9 above)
Comment (krystian14d) [reply src/test/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletControllerTest.java:64]: Updated.

File: src/main/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletController.java (hunk 10 above)
Comment (krystian14d) [reply src/main/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletController.java:32]: Updated.

File: src/main/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletController.java (hunk 11 above)
Comment (krystian14d) [reply src/main/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletController.java:30]: That is a good question. To be honest in the task descripion it is not clearly stated how the endpoint should look like, therefore I decided to use just "/update" path without ID, because ID is included in the response body, simultaneously having the validation of provided parameters.

Comment (Iwona007) [reply src/main/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletController.java:33]: Have you ever heard about "Richardson Maturity Model"? We implement Rest Architecture and this mean that we use "Richardson Maturity Model". For our project we use level 2: https://martinfowler.com/articles/richardsonMaturityModel.html#level2 . According to this cannot use verbs as an endpoint  (update is a verb). Please remove "/update". Because we have @RequestMapping("api/wallet") - for the method "updateWallet" you can use {id} as as a @PathVariable.

File: src/test/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletControllerTest.java (hunk 9 above)
Comment (Iwona007) [reply src/test/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletControllerTest.java]: Have you ever heard about "Richardson Maturity Model"? We implement Rest Architecture and this mean that we use "Richardson Maturity Model". For our project we use level 2: https://martinfowler.com/articles/richardsonMaturityModel.html#level2 . According to this cannot use verbs as an endpoint (edit is a verb). Please remove "/edit".  Additionally please read more about "Richardson Maturity Model"

File: src/test/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletControllerTest.java [hunk 12]
```diff
@@ -1,7 +1,237 @@
...
+                        .contentType(MediaType.APPLICATION_JSON)
+                        .content(Objects.requireNonNull(objectToJson(editWalletDto)))
+        );
+
+        //THEN
+        editResultActions.andExpect(status().isBadRequest());
+    }
+
+    @Test
+    void itShouldReturnBadRequestWhenIdIsZero() throws Exception {
```
Comment (dawciobiel) [inline src/test/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletControllerTest.java:167]: Test should fail when `id` is lower than '1'. Value is lower than `1` mean it's `0` or even negative value.

File: src/main/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletController.java [hunk 13]
```diff
@@ -14,4 +29,21 @@
...
+
+        WalletDTO walletDTO = walletService.updateWallet(updateWalletDto);
+        HttpHeaders headers = new HttpHeaders();
+        headers.add("message", "You have successfully updated Wallet!");
+        return new ResponseEntity<>(walletDTO, headers, HttpStatus.OK);
+    }
+
+    @ResponseStatus(HttpStatus.BAD_REQUEST)
+    @ExceptionHandler(MethodArgumentNotValidException.class)
+    Map<String, String> handleMethodArgumentNotValidException(MethodArgumentNotValidException ex) {
```
Comment (dawciobiel) [inline src/main/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletController.java:44]: Why this method is not inside .. i.e: `WalletExceptionHandler` class?

[**Almost** good implemented `WalletExceptionHandler`](
https://github.com/bycza-zagroda/track-expenses-app-backend/blob/9c42da340b00f8f9c2d380557014114007aed9b4/src/main/java/pl/byczazagroda/trackexpensesappbackend/exception/WalletExceptionHandler.java)

**Review by dawciobiel — CHANGES_REQUESTED** [review]

File: src/test/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletControllerTest.java (hunk 12 above)
Comment (krystian14d) [reply src/test/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletControllerTest.java:167]: I propose to add one more test, where ID with negative value will be tested. Is that okay?

File: src/main/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletController.java (hunk 13 above)
Comment (dawciobiel) [reply src/main/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletController.java:44]: Nope. `GlobalExceptionHandler..` is to catch for example:

-     SQLException
-     SQLNonTransientException
-     SQLTransientException
-     SQLRecoverableException
-     etc..

Comment (krystian14d) [reply src/main/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletController.java]: OK, added WalletExceptionHandler class.

File: src/test/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletControllerTest.java (hunk 12 above)
Comment (krystian14d) [reply src/test/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletControllerTest.java:167]: Decided to add new test to check when ID will be negative.

File: src/main/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletController.java (hunk 10 above)
Comment (Iwona007) [reply src/main/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletController.java:30]: Please remove the "/update". this the same: https://github.com/bycza-zagroda/track-expenses-app-backend/pull/53#discussion_r996501486

File: src/test/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletControllerTest.java [hunk 14]
```diff
@@ -186,6 +186,29 @@ void itShouldReturnBadRequestWhenIdIsZero() throws Exception {
...
+                new WalletDTO(
+                        Mockito.anyLong(),
+                        Mockito.anyString(),
+                        Instant.now()
+                )
+        );
+
+        //WHEN
+        ResultActions editResultActions = mockMvc.perform(
+                patch("/api/wallet")
```
Comment (pyanik) [inline src/test/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletControllerTest.java:203]: Why `patch()` method is used for testing, while in `WalletController.java` only Put REST API method is defined?

Did you run those tests cases? Becasue i'm pretty sure that they failed.

File: src/main/java/pl/byczazagroda/trackexpensesappbackend/service/WalletServiceImpl.java [hunk 15]
```diff
@@ -1,13 +1,36 @@
...
+import pl.byczazagroda.trackexpensesappbackend.dto.UpdateWalletDTO;
+import pl.byczazagroda.trackexpensesappbackend.dto.WalletDTO;
+import pl.byczazagroda.trackexpensesappbackend.dto.WalletModelMapper;
+import pl.byczazagroda.trackexpensesappbackend.exception.ResourceNotFoundException;
+import pl.byczazagroda.trackexpensesappbackend.model.Wallet;
 import pl.byczazagroda.trackexpensesappbackend.repository.WalletRepository;
 
+import javax.transaction.Transactional;
+
+@Primary
```
Comment (pyanik) [inline src/main/java/pl/byczazagroda/trackexpensesappbackend/service/WalletServiceImpl.java:15]: Why `@Primary` annotation is used. Do we have more `WalletService` implementations?

File: src/main/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletController.java [hunk 16]
```diff
@@ -14,4 +23,15 @@
 public class WalletController {
 
     private final WalletService walletService;
+
+    @PutMapping
```
Comment (pyanik) [inline src/main/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletController.java:27]: I suggest to include `id` of the wallet to update in the endpoint. In that case we have to add second argument @Pathvariable Long id to the `updateWallet()` method and replace `@RequestBody @Valid UpdateWalletDTO updateWalletDto` with the object of existing `CreateWalletDTO.class` - In that case we can rename this class to `InputWalletDTO.class`

File: src/main/java/pl/byczazagroda/trackexpensesappbackend/exception/GlobalExceptionHandlerController.java [hunk 17]
```diff
@@ -44,10 +44,18 @@ public String handleMethodArgumentNotValidException(MethodArgumentNotValidExcept
         return e.getMessage();
     }
 
+
     @ExceptionHandler(ResourceNotFoundException.class)
     @ResponseStatus(HttpStatus.INTERNAL_SERVER_ERROR)
```
Comment (dawciobiel) [inline src/main/java/pl/byczazagroda/trackexpensesappbackend/exception/GlobalExceptionHandlerController.java:49]: There should be some HttpStatus from 4xx series. Maybe 404 will be proper.
Because HttpStatus.INTERNAL_SERVER_ERROR 500 mean:
> The server encountered an unexpected condition that prevented it from fulfilling the request.

https://www.webfx.com/web-development/glossary/http-status-codes/what-is-a-500-status-code/

File: src/main/java/pl/byczazagroda/trackexpensesappbackend/service/WalletServiceImpl.java [hunk 18]
```diff
@@ -29,7 +32,19 @@ public WalletDTO updateWallet(UpdateWalletDTO dto) throws ResourceNotFoundExcept
...
+    public WalletDTO createWallet(CreateWalletDTO createWalletDTO) {
+        String walletName = createWalletDTO.name();
+        Wallet wallet = new Wallet(walletName);
+        Wallet savedWallet = walletRepository.save(wallet);
+        boolean isWalletExists = walletRepository.existsById(savedWallet.getId());
+
+        if (isWalletExists) {
+            return walletModelMapper.mapWalletEntityToWalletDTO(savedWallet);
+        }
+        throw new ResourceNotSavedException("Sorry. Something went wrong and your Wallet is not saved. Contact administrator.");
```
Comment (dawciobiel) [inline src/main/java/pl/byczazagroda/trackexpensesappbackend/service/WalletServiceImpl.java:48]: Please change it to: "Sorry. Something went wrong and your Wallet is not saved. **Please** contact administrator."

File: src/test/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletControllerTest.java [hunk 19]
```diff
@@ -107,6 +112,22 @@ void itShouldReturnBadRequestWhenNameIsTooLong() throws Exception {
...
+        CreateWalletDTO createWalletDTO = new CreateWalletDTO("");
+        given(walletService.createWallet(createWalletDTO))
+                .willReturn(new WalletDTO(1L, "", Instant.now()));
+
+        // when
+        ResultActions result = mockMvc.perform(post("/api/wallet")
+                .contentType(MediaType.APPLICATION_JSON)
+                .content(Objects.requireNonNull(objectMapper.writeValueAsString(createWalletDTO))));
+        // then
+
```
Comment (dawciobiel) [inline src/test/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletControllerTest.java:127]: Empty line #127 not needed

File: src/test/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletControllerTest.java [hunk 20]
```diff
@@ -72,6 +163,26 @@ void shouldThrowAnExceptionWhenNameIsNull() throws Exception {
...
+    @Test
+    void itShouldReturnBadRequestWhenIdIsNull() throws Exception {
+        //GIVEN
+        Instant timeCreated = Instant.now();
+
+        UpdateWalletDTO updateWalletDTO = new UpdateWalletDTO(null, "@#$%^&");
+        given(walletService.updateWallet(updateWalletDTO))
+                .willReturn(new WalletDTO(null, "", timeCreated));
+
+        //WHEN
```
Comment (MateuszBednarczyk) [inline src/test/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletControllerTest.java:175]: Why you wrote it in upper case?

File: src/test/java/pl/byczazagroda/trackexpensesappbackend/service/WalletServiceImplTest.java [hunk 21]
```diff
@@ -64,7 +72,115 @@ void itShouldThrowWhenWalletNotFound() {
...
-        assertThatThrownBy(() -> underTest.updateWallet(updateWalletDto))
+        assertThatThrownBy(() -> walletService.updateWallet(updateWalletDto))
                 .isInstanceOf(ResourceNotFoundException.class);
     }
-}
\ No newline at end of file
+
+    @Test
+    void shouldCreateWalletProperly() {
+//given
```
Comment (dawciobiel) [inline src/test/java/pl/byczazagroda/trackexpensesappbackend/service/WalletServiceImplTest.java:81]: Please do comments formatting in same way for:
// given
// when
// then

File: src/main/java/pl/byczazagroda/trackexpensesappbackend/service/WalletServiceImpl.java [hunk 22]
```diff
@@ -18,6 +22,18 @@ public class WalletServiceImpl implements WalletService {
     private final WalletRepository walletRepository;
     private final WalletModelMapper walletModelMapper;
 
+    @Override
+    @Transactional
+    public WalletDTO updateWallet(UpdateWalletDTO dto) throws ResourceNotFoundException {
+        Wallet wallet = walletRepository.findById(dto.id()).orElseThrow(() ->
+        {
```
Comment (MateuszBednarczyk) [inline src/main/java/pl/byczazagroda/trackexpensesappbackend/service/WalletServiceImpl.java:29]: I think, that this bracket should be in the previous line

File: src/main/java/pl/byczazagroda/trackexpensesappbackend/service/WalletServiceImpl.java [hunk 23]
```diff
@@ -18,6 +22,18 @@ public class WalletServiceImpl implements WalletService {
...
+            throw new ResourceNotFoundException(String.format("Wallet with given ID: %s does not exist", dto.id()));
+        });
+
```
Comment (MateuszBednarczyk) [inline src/main/java/pl/byczazagroda/trackexpensesappbackend/service/WalletServiceImpl.java:32]: Why there's an empty line?

File: src/test/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletControllerTest.java [hunk 24]
```diff
@@ -41,6 +45,72 @@ class WalletControllerTest {
...
+        editResultActions.andExpect(status().isOk())
+                .andExpect(jsonPath("$.id", Matchers.equalTo(updateWalletDto.id().intValue())))
+                .andExpect(jsonPath("$.name", Matchers.equalTo(updateWalletDto.name())));
+    }
+
+    @Test
+    void itShouldReturnBadRequestWhenNameIsEmpty() throws Exception {
+        //GIVEN
+        Instant timeCreated = Instant.now();
+
```
Comment (MateuszBednarczyk) [inline src/test/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletControllerTest.java:79]: Why there's an empty line?

**Review by dawciobiel — CHANGES_REQUESTED** [review]
Changes required

File: src/test/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletControllerTest.java [hunk 25]
```diff
@@ -89,6 +200,45 @@ void shouldThrowAnExceptionWhenNameIsTooLong() throws Exception {
...
+        given(walletService.updateWallet(updateWalletDTO))
+                .willReturn(new WalletDTO(-1L, "", timeCreated));
+
+        //WHEN
+        ResultActions result = mockMvc.perform(
+                put("/api/wallet")
+                        .contentType(MediaType.APPLICATION_JSON)
+                        .content(Objects.requireNonNull(objectMapper.writeValueAsString(updateWalletDTO)))
+        );
+
```
Comment (MateuszBednarczyk) [inline src/test/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletControllerTest.java:238]: Why you didn't describe `//then` section?

**Review by MateuszBednarczyk — CHANGES_REQUESTED** [review]

File: src/main/java/pl/byczazagroda/trackexpensesappbackend/exception/GlobalExceptionHandlerController.java (hunk 17 above)
Comment (krystian14d) [reply src/main/java/pl/byczazagroda/trackexpensesappbackend/exception/GlobalExceptionHandlerController.java:49]: Done - changed to 404.

File: src/main/java/pl/byczazagroda/trackexpensesappbackend/service/WalletServiceImpl.java (hunk 18 above)
Comment (krystian14d) [reply src/main/java/pl/byczazagroda/trackexpensesappbackend/service/WalletServiceImpl.java:48]: Done.

File: src/test/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletControllerTest.java (hunk 19 above)
Comment (krystian14d) [reply src/test/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletControllerTest.java:127]: Empty line removed.

File: src/test/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletControllerTest.java (hunk 20 above)
Comment (krystian14d) [reply src/test/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletControllerTest.java:175]: TO have better overview and orientation, in case if test method will be too long? Is that upper case an issue? Please let me know whether it should be upper or lower case convention.

File: src/main/java/pl/byczazagroda/trackexpensesappbackend/service/WalletServiceImpl.java (hunk 22 above)
Comment (krystian14d) [reply src/main/java/pl/byczazagroda/trackexpensesappbackend/service/WalletServiceImpl.java:29]: Done.

File: src/main/java/pl/byczazagroda/trackexpensesappbackend/service/WalletServiceImpl.java (hunk 23 above)
Comment (krystian14d) [reply src/main/java/pl/byczazagroda/trackexpensesappbackend/service/WalletServiceImpl.java:32]: Empty line removed.

File: src/test/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletControllerTest.java (hunk 24 above)
Comment (krystian14d) [reply src/test/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletControllerTest.java:79]: To have kind of 'sectioning' in the testing method. But line has been removed if that is a case.

File: src/test/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletControllerTest.java (hunk 25 above)
Comment (krystian14d) [reply src/test/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletControllerTest.java:238]: Probably disappeared during merging with develop branch. '//then' section added.

File: src/test/java/pl/byczazagroda/trackexpensesappbackend/service/WalletServiceImplTest.java (hunk 21 above)
Comment (krystian14d) [reply src/test/java/pl/byczazagroda/trackexpensesappbackend/service/WalletServiceImplTest.java:81]: Formatting of comments refactored.

File: src/test/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletControllerTest.java (hunk 20 above)
Comment (krystian14d) [reply src/test/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletControllerTest.java:175]: changed to lower case.

File: src/test/java/pl/byczazagroda/trackexpensesappbackend/service/WalletServiceImplTest.java [hunk 26]
```diff
@@ -38,10 +46,42 @@ class WalletServiceImplTest {
...
 
+    @Test
+    void itShouldUpdateWalletName() {
+        // given
+        UpdateWalletDTO updateWalletDto = new UpdateWalletDTO(1L, "walletName");
+        Wallet wallet = new Wallet("anyName");
+        wallet.setId(1L);
+        Instant time = Instant.now();
+        wallet.setCreationDate(Instant.now());
+
```
Comment (MateuszBednarczyk) [inline src/test/java/pl/byczazagroda/trackexpensesappbackend/service/WalletServiceImplTest.java:57]: why is there an empty line?

File: src/test/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletControllerTest.java [hunk 27]
```diff
@@ -89,6 +198,46 @@ void shouldThrowAnExceptionWhenNameIsTooLong() throws Exception {
...
+        // then
+        result.andExpect(status().isBadRequest());
+    }
+
+    @Test
+    void itShouldReturnBadRequestWhenIdIsNegative() throws Exception {
+        // given
+        Instant timeCreated = Instant.now();
+
```
Comment (MateuszBednarczyk) [inline src/test/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletControllerTest.java:225]: why is there an empty line?

File: src/test/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletControllerTest.java [hunk 28]
```diff
@@ -41,6 +45,71 @@ class WalletControllerTest {
...
+        // then
+        result.andExpect(status().isBadRequest());
+    }
+
+    @Test
+    void itShouldReturnBadRequestWhenNameIsTooLong() throws Exception {
+        // given
+        Instant timeCreated = Instant.now();
+
```
Comment (MateuszBednarczyk) [inline src/test/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletControllerTest.java:98]: why is there an empty line?

File: src/test/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletControllerTest.java [hunk 29]
```diff
@@ -41,6 +45,71 @@ class WalletControllerTest {
     @Autowired
     private ObjectMapper objectMapper;
 
+    @Test
+    void itShouldReturnStatusOKAndCorrectResponseBody() throws Exception {
+        //GIVEN
+        Instant timeCreated = Instant.now();
+
```
Comment (MateuszBednarczyk) [inline src/test/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletControllerTest.java:52]: why is there an empty line?

File: src/test/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletControllerTest.java [hunk 30]
```diff
@@ -41,6 +45,71 @@ class WalletControllerTest {
     @Autowired
     private ObjectMapper objectMapper;
 
+    @Test
+    void itShouldReturnStatusOKAndCorrectResponseBody() throws Exception {
+        //GIVEN
```
Comment (MateuszBednarczyk) [inline src/test/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletControllerTest.java:50]: please, describe given section with lowercase.

File: src/test/java/pl/byczazagroda/trackexpensesappbackend/service/WalletServiceImplTest.java (hunk 26 above)
Comment (krystian14d) [reply src/test/java/pl/byczazagroda/trackexpensesappbackend/service/WalletServiceImplTest.java:57]: updated.

File: src/test/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletControllerTest.java (hunk 27 above)
Comment (krystian14d) [reply src/test/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletControllerTest.java:225]: updated.

File: src/test/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletControllerTest.java (hunk 29 above)
Comment (krystian14d) [reply src/test/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletControllerTest.java:52]: updated

File: src/test/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletControllerTest.java (hunk 30 above)
Comment (krystian14d) [reply src/test/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletControllerTest.java:50]: updated.

**Review by MateuszBednarczyk — APPROVED** [review]

**Review by dawciobiel — APPROVED** [review]
Looks good for me

File: src/main/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletController.java [hunk 31]
```diff
@@ -24,6 +26,17 @@ public class WalletController {
 
     private final WalletService walletService;
 
+
+    @PutMapping
+    public ResponseEntity<WalletDTO> updateWallet(@RequestBody @Valid UpdateWalletDTO updateWalletDto) {
```
[the remark this answers is not in the document]
Comment (dawciobiel) [reply src/main/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletController.java]: To be honest, following the order of letters in the Latin alphabet, first **R**equestBody, and then **V**alid
The question, however, is which of the annotations will be launched first.
It is true - it does not matter here.

**Alphabet**
[Link](https://en.wikipedia.org/wiki/Latin_script)

Regards.

**Review by dawciobiel — APPROVED** [review]
Comments resolved.
Looks good for me.

File: src/main/java/pl/byczazagroda/trackexpensesappbackend/dto/WalletModelMapper.java [hunk 32]
```diff
@@ -0,0 +1,11 @@
+package pl.byczazagroda.trackexpensesappbackend.dto;
+
+import org.mapstruct.InjectionStrategy;
+import org.mapstruct.Mapper;
+import pl.byczazagroda.trackexpensesappbackend.model.Wallet;
+
+@Mapper(componentModel = "spring")
+public interface WalletModelMapper {
```
Comment (MateuszBednarczyk) [inline src/main/java/pl/byczazagroda/trackexpensesappbackend/dto/WalletModelMapper.java:8]: Please, delete this interface, and next time make sure that you are up to date with develop branch, because ``WalletModelMapper`` already exists in ``mapper`` package. So please, next time take a note of it.

**Review by MateuszBednarczyk — CHANGES_REQUESTED** [review]

**Review by MateuszBednarczyk — APPROVED** [review]

**Review by dawciobiel — APPROVED** [review]
Looks good for me
